### PR TITLE
For loop for arrays

### PR DESCRIPTION
**Pull request by pmqtt**
Hello folks,
I have the for-loop statement implements... Test Cases are  in testdata/for/

Comments and suggestions for improvement are very welcome, because my implementation manipulate the AST to transform the for-loop in the aquivalent while form.

I hope it is useful und best regrads!

### CODE REVIEW

File: explorer/ast/statement.h [hunk 1]
```diff
@@ -298,6 +298,108 @@ class While : public Statement {
...
+      body_stmts.push_back(iter);
+    };
+    body_stmts.push_back(std::move(increment_index_stmt));
+
+    auto block = arena->New<Block>(
+        source_loc, std::vector<Nonnull<Statement*>>({body_stmts}));
+    auto while_stmt = arena->New<While>(source_loc, condition, block);
+
+    statements_.push_back(std::move(while_stmt));
+  }
```
Comment (geoffromer) [inline explorer/ast/statement.h]: According to our [project style guide](https://github.com/carbon-language/carbon-lang/blob/trunk/docs/project/cpp_style_guide.md), functions should only be inlined when they are [very short](https://google.github.io/styleguide/cppguide.html#Inline_Functions). I think addressing my comment above might make this function body very short, but if not, please move it to `statement.cpp`.

File: explorer/ast/statement.h [hunk 2]
```diff
@@ -298,6 +298,108 @@ class While : public Statement {
...
 
+class For : public Statement {
+ public:
+  For(SourceLocation source_loc, Nonnull<BindingPattern*> variable_declaration,
+      Nonnull<Expression*> loop_target, Nonnull<Block*> body)
+      : Statement(AstNodeKind::For, source_loc),
+        variable_declaration_(variable_declaration),
+        loop_target_(loop_target),
+        body_(body) {
+    Nonnull<Arena*> arena = new Arena();
```
Comment (geoffromer) [inline explorer/ast/statement.h]: This will leak `arena` at the end of the function. If this constructor needs to allocate objects on an arena, it should take the arena as an in/out parameter. However, I'm not sure it should be allocating anything -- see my next comment.

File: explorer/ast/statement.h [hunk 3]
```diff
@@ -298,6 +298,108 @@ class While : public Statement {
...
+    auto index_name =
+        arena->New<IdentifierExpression>(source_loc, "$yy_forloop");
```
Comment (geoffromer) [inline explorer/ast/statement.h]: We generally try to have the parsed AST match the syntax that the user supplied, because otherwise things like error messages and trace output can get very confusing, because they're talking about code that the user can't see. We've sometimes talked about implementing some features in terms of AST rewriting, but haven't figured out a design that we're happy with.

Instead of introducing these artificial AST nodes so that they can be fed to the interpreter, my recommendation would be to have logic in the part of the interpreter that implements `For` that directly does what execution of those nodes would do. For example, instead of synthesizing an AST node that calls `__intrinsic_array_sz`, have code in the interpreter that directly calls `.elements().vec().size()` on the array (as a side benefit, I think that means we don't need `__intrinsiz_array_sz` in the first place). Does that make sense?

File: explorer/interpreter/type_checker.cpp [hunk 4]
```diff
@@ -2223,6 +2223,20 @@ auto TypeChecker::TypeCheckExp(Nonnull<Expression*> e,
           e->set_value_category(ValueCategory::Let);
           return Success();
         }
+        case IntrinsicExpression::Intrinsic::ArraySz: {
+          if (args.size() != 1) {
+            return CompilationError(e->source_loc())
+                   << "__intrinsic_array_sz takes 1 argument, received "
+                   << args.size();
+          }
+          /*CARBON_RETURN_IF_ERROR(ExpectExactType(
```
Comment (geoffromer) [inline explorer/interpreter/type_checker.cpp]: Did you mean to comment this out? If so, why?

File: explorer/testdata/for/for-loop.carbon [hunk 5]
```diff
@@ -0,0 +1,52 @@
+// Part of the Carbon Language project, under the Apache License v2.0 with LLVM
```
Comment (geoffromer) [inline explorer/testdata/for/for-loop.carbon]: We should also have some tests that cover `break` and `continue`

File: explorer/testdata/for/for-loop.carbon [hunk 6]
```diff
@@ -0,0 +1,52 @@
...
+// CHECK: result: 8
+
+package ExplorerTest api;
+
+
+
+
+fn Main() -> i32 {
+  var ar: [i32; 4] = (0, 1,2,3);
+  var i : i32 = __intrinsic_array_size(ar);
```
Comment (geoffromer) [inline explorer/testdata/for/for-loop.carbon]: We generally try to avoid using intrinsics in user code, and it looks like `i` isn't used anyway. Can we remove this line?

File: explorer/testdata/for/for-loop.carbon [hunk 7]
```diff
@@ -0,0 +1,52 @@
...
+       }
+       if( x == 1){
+        Print("HALLO WELT 1");
+       }
+       if( x == 2){
+        Print("HALLO WELT 2");
+       }
+       if( x == 3){
+        Print("HALLO WELT 3");
+       }
```
Comment (geoffromer) [inline explorer/testdata/for/for-loop.carbon]: Would something like this work instead?

```suggestion
      Print("HALLO WELT ", x);
```

If so, that would probably be clearer.

**Comment by pmqtt** [comment]
Thanks for the review... I should think again, how to implement this...

File: explorer/ast/statement.h (hunk 3 above)
Comment (pmqtt) [reply explorer/ast/statement.h]: Ok! I will remove the code and try to implement as logic in the node.

File: explorer/ast/expression.cpp [hunk 8]
```diff
@@ -184,6 +185,9 @@ void Expression::Print(llvm::raw_ostream& out) const {
         case IntrinsicExpression::Intrinsic::Dealloc:
           out << "delete";
           break;
+        case IntrinsicExpression::Intrinsic::ArraySz:
+          out << "array_sz";
```
Comment (zygoloid) [inline explorer/ast/expression.cpp]: This isn't consistent with the spelling you used earlier.

File: explorer/ast/statement.cpp [hunk 9]
```diff
@@ -42,6 +42,12 @@ void Statement::PrintDepth(int depth, llvm::raw_ostream& out) const {
       while_stmt.body().PrintDepth(depth - 1, out);
       break;
     }
+    case StatementKind::For: {
+      const auto& for_stmt = cast<For>(*this);
+      out << "for (" << for_stmt.loop_target() << ")\n";
```
Comment (zygoloid) [inline explorer/ast/statement.cpp]:
```suggestion
      out << "for (" << for_stmt.variable_declaration() << " in " << for_stmt.loop_target() << ")\n";
```

File: explorer/ast/statement.h (hunk 3 above)
Comment (zygoloid) [reply explorer/ast/statement.h]: We may need or want to create additional AST nodes to represent the semantics of the  loop -- for example, if we have an interface for converting an arbitrary sequence into something we can iterate over. If you find that's necessary, that would be better done in type_checking.cpp when we type-check the statement, rather than here.

**Comment by pmqtt** [comment]
Hello folks, I have reimplement the for-loop statement.

File: explorer/ast/expression.cpp (hunk 8 above)
Comment (pmqtt) [reply explorer/ast/expression.cpp]: It's removed!

**Comment by Pixep** [comment]
Relates to #1592

File: explorer/testdata/for/for-loop.carbon (hunk 5 above)
Comment (clavin) [reply explorer/testdata/for/for-loop.carbon]: On the same topic, it would also be nice to have a test with an empty array

File: explorer/testdata/for/for-loop.carbon (hunk 6 above)
Comment (pmqtt) [reply explorer/testdata/for/for-loop.carbon]: removed

File: explorer/testdata/for/for-loop.carbon (hunk 5 above)
Comment (pmqtt) [reply explorer/testdata/for/for-loop.carbon]: tests for break and continue are added

File: explorer/interpreter/type_checker.cpp (hunk 4 above)
Comment (pmqtt) [reply explorer/interpreter/type_checker.cpp]: type checking is corrected

File: explorer/interpreter/interpreter.cpp [hunk 10]
```diff
@@ -1469,6 +1469,68 @@ auto Interpreter::StepStmt() -> ErrorOr<Success> {
...
+          CARBON_ASSIGN_OR_RETURN(
+              Nonnull<const Value*> assigned_array_element,
+              todo_.ValueOfNode(*(loop_var->value_node()), stmt.source_loc()));
+
+          if (const auto* lvalue = dyn_cast<LValue>(assigned_array_element)) {
+            CARBON_RETURN_IF_ERROR(heap_.Write(
+                lvalue->address(), source_array->elements()[current_index],
+                stmt.source_loc()));
+          } else {
+            llvm::outs() << "Failure in " << __LINE__ << "\n";
```
Comment (zygoloid) [inline explorer/interpreter/interpreter.cpp]: Please use `CARBON_CHECK` or `CARBON_FATAL` as appropriate here.

File: explorer/interpreter/type_checker.cpp [hunk 11]
```diff
@@ -2863,6 +2863,24 @@ auto TypeChecker::TypeCheckStmt(Nonnull<Statement*> s,
       CARBON_RETURN_IF_ERROR(TypeCheckStmt(&while_stmt.body(), impl_scope));
       return Success();
     }
+    case StatementKind::For: {
+      auto& for_stmt = cast<For>(*s);
+      ImplScope inner_impl_scope;
+      inner_impl_scope.AddParent(&impl_scope);
```
Comment (zygoloid) [inline explorer/interpreter/type_checker.cpp:2955]: I don't think we should need a separate `ImplScope` here, since `impl`s can't be nested within the `for` statement.

File: explorer/interpreter/type_checker.cpp [hunk 12]
```diff
@@ -2863,6 +2863,24 @@ auto TypeChecker::TypeCheckStmt(Nonnull<Statement*> s,
...
+      CARBON_RETURN_IF_ERROR(TypeCheckPattern(&for_stmt.variable_declaration(),
+                                              std::nullopt, inner_impl_scope,
+                                              ValueCategory::Let));
```
Comment (zygoloid) [inline explorer/interpreter/type_checker.cpp]: If we type-check the range first, we can pass in an expected type here, which should allow us to support things like `for (x: auto in arr)`.

File: explorer/interpreter/type_checker.cpp [hunk 13]
```diff
@@ -2863,6 +2863,24 @@ auto TypeChecker::TypeCheckStmt(Nonnull<Statement*> s,
...
+      CARBON_RETURN_IF_ERROR(
+          TypeCheckExp(&for_stmt.loop_target(), inner_impl_scope));
+
+      CARBON_RETURN_IF_ERROR(
+          ImplicitlyConvert("declaration of for loop variable", impl_scope,
+                            &for_stmt.loop_target(),
+                            &for_stmt.variable_declaration().static_type()));
```
Comment (zygoloid) [inline explorer/interpreter/type_checker.cpp]: Hm, I would have expected that we want to convert the array element type of the type of the loop target, not the type of the loop target itself, here.

File: explorer/interpreter/type_checker.cpp (hunk 11 above)
Comment (pmqtt) [reply explorer/interpreter/type_checker.cpp:2955]: Please corrent me, but isn't it so:

```
Block
for( x: i32 in ar) //new Block{
//new Block
}
Block
```
Because after the execution of the loop, *x* shouldn't be found in Block.

Comment (pmqtt) [reply explorer/interpreter/type_checker.cpp:2955]: Another problem is  that TypeCheckPattern expected a non const ImplScope,  but impl_scope is const!

File: explorer/interpreter/type_checker.cpp (hunk 13 above)
Comment (pmqtt) [reply explorer/interpreter/type_checker.cpp]: thanks, solved!

File: explorer/interpreter/type_checker.cpp (hunk 12 above)
Comment (pmqtt) [reply explorer/interpreter/type_checker.cpp]: thanks, solved!

File: explorer/interpreter/interpreter.cpp (hunk 10 above)
Comment (pmqtt) [reply explorer/interpreter/interpreter.cpp]: Replaced!

File: explorer/interpreter/interpreter.cpp [hunk 14]
```diff
@@ -1469,6 +1469,68 @@ auto Interpreter::StepStmt() -> ErrorOr<Success> {
         }
       }
     }
+    case StatementKind::For: {
+      if (act.pos() == 0) {
+        act.AddResult(arena_->New<IntValue>(0));
```
Comment (geoffromer) [inline explorer/interpreter/interpreter.cpp]: Would it make sense to move this to the `act.pos() == 1` case?

File: explorer/interpreter/interpreter.cpp [hunk 15]
```diff
@@ -1469,6 +1469,68 @@ auto Interpreter::StepStmt() -> ErrorOr<Success> {
...
+        return todo_.Spawn(
+            std::make_unique<ExpressionAction>(&cast<For>(stmt).loop_target()));
+      }
+      if (act.pos() == 1) {
+        Nonnull<const TupleValue*> source_array =
+            cast<const TupleValue>(act.results().back());
+
+        auto end_index = static_cast<int>(source_array->elements().size());
+        auto start_index = cast<IntValue>(act.results()[0])->value();
```
Comment (geoffromer) [inline explorer/interpreter/interpreter.cpp]: Isn't this always going to be 0?

File: explorer/interpreter/interpreter.cpp [hunk 16]
```diff
@@ -1469,6 +1469,68 @@ auto Interpreter::StepStmt() -> ErrorOr<Success> {
...
+          return todo_.Spawn(std::make_unique<PatternAction>(
+              &cast<For>(stmt).variable_declaration()));
+        }
+        return todo_.FinishAction();
+      }
+      if (act.pos() == 2) {
+        Nonnull<const BindingPlaceholderValue*> loop_var =
+            cast<const BindingPlaceholderValue>(act.results().back());
+        Nonnull<const TupleValue*> source_array =
+            cast<const TupleValue>(act.results()[1]);
```
Comment (geoffromer) [inline explorer/interpreter/interpreter.cpp]: It's kind of hard to keep track of which result index has which meaning. Can we define and use named constants instead of 0/1/2 magic numbers?

File: explorer/interpreter/interpreter.cpp [hunk 17]
```diff
@@ -1503,6 +1503,68 @@ auto Interpreter::StepStmt() -> ErrorOr<Success> {
...
+
+          if (const auto* lvalue = dyn_cast<LValue>(assigned_array_element)) {
+            CARBON_RETURN_IF_ERROR(heap_.Write(
+                lvalue->address(), source_array->elements()[current_index],
+                stmt.source_loc()));
+          } else {
+            CARBON_FATAL();
+          }
+
+          act.ReplaceResult(0, arena_->New<IntValue>(current_index + 1));
```
Comment (geoffromer) [inline explorer/interpreter/interpreter.cpp]: I like this approach, but for consistency I think we should handle `While` the same way.

File: explorer/interpreter/interpreter.cpp [hunk 18]
```diff
@@ -1503,6 +1503,68 @@ auto Interpreter::StepStmt() -> ErrorOr<Success> {
...
+          Nonnull<const TupleValue*> source_array =
+              cast<const TupleValue>(act.results()[1]);
+          Nonnull<const BindingPlaceholderValue*> loop_var =
+              cast<const BindingPlaceholderValue>(act.results()[3]);
+
+          CARBON_ASSIGN_OR_RETURN(
+              Nonnull<const Value*> assigned_array_element,
+              todo_.ValueOfNode(*(loop_var->value_node()), stmt.source_loc()));
+
+          if (const auto* lvalue = dyn_cast<LValue>(assigned_array_element)) {
```
Comment (geoffromer) [inline explorer/interpreter/interpreter.cpp]: It seems like this should be `cast` rather than `dyn_cast` -- that way we get the fatal error on failure automatically.

File: explorer/interpreter/resolve_names.cpp [hunk 19]
```diff
@@ -413,6 +413,17 @@ static auto ResolveNames(Statement& statement, StaticScope& enclosing_scope)
       CARBON_RETURN_IF_ERROR(ResolveNames(while_stmt.body(), enclosing_scope));
       break;
     }
+    case StatementKind::For: {
+      StaticScope block_scope;
```
Comment (geoffromer) [inline explorer/interpreter/resolve_names.cpp]: This scope doesn't really correspond to a block, right? It corresponds to the whole `for` statement. That being the case, maybe this should be called something like `statement_scope`?

File: explorer/testdata/for/for-loop-auto.carbon [hunk 20]
```diff
@@ -0,0 +1,25 @@
+// Part of the Carbon Language project, under the Apache License v2.0 with LLVM
+// Exceptions. See /LICENSE for license information.
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+//
+// RUN: %{explorer} %s 2>&1 | \
+// RUN:   %{FileCheck} --match-full-lines --allow-unused-prefixes=false %s
+// RUN: %{explorer} --parser_debug --trace_file=- %s 2>&1 | \
+// RUN:   %{FileCheck} --match-full-lines --allow-unused-prefixes %s
+// AUTOUPDATE: %{explorer} %s
```
Comment (geoffromer) [inline explorer/testdata/for/for-loop-auto.carbon]: Have you run `explorer/update_checks.py`? I think it will add `CHECK` lines for the `Print` output; if so, we should add them now.

File: explorer/testdata/for/for-loop-auto.carbon [hunk 21]
```diff
@@ -0,0 +1,25 @@
...
+
+
+fn Main() -> i32 {
+  var ar: [i32; 4] = (0, 1,2,3);
+  var count : i32 = 0;
+  for( x: i32 in ar){
+      Print("HALLO WELT ", x);
+      count = count +1;
+  }
+  return 4;
```
Comment (geoffromer) [inline explorer/testdata/for/for-loop-auto.carbon]: Why is this 4 rather than 0 or something like `count`?

File: explorer/testdata/for/for-loop-sum.carbon [hunk 22]
```diff
@@ -0,0 +1,24 @@
+// Part of the Carbon Language project, under the Apache License v2.0 with LLVM
```
Comment (geoffromer) [inline explorer/testdata/for/for-loop-sum.carbon]: Does this test anything interesting about `for` support that isn't covered by `for-loop.carbon`?

File: explorer/testdata/for/for-loop-auto.carbon [hunk 23]
```diff
@@ -0,0 +1,25 @@
+// Part of the Carbon Language project, under the Apache License v2.0 with LLVM
```
Comment (geoffromer) [inline explorer/testdata/for/for-loop-auto.carbon]: What is this file supposed to test? The file name mentions `auto`, but I don't see `auto` anywhere in the code.

Comment (geoffromer) [inline explorer/testdata/for/for-loop-auto.carbon]: For consistency with the other `.carbon` files, I think these files' names should use `_` instead of `-`.

File: explorer/testdata/for/for-loop.carbon (hunk 5 above)
Comment (geoffromer) [reply explorer/testdata/for/for-loop.carbon]: I agree with @clavin that we should also test iterating over an empty array.

File: explorer/data/prelude.carbon [hunk 24]
```diff
@@ -141,23 +141,23 @@ impl String as CompareWith(Self) {
 }
 
 interface LessWith(U:! Type) {
-  fn Less[me: Self](other: U) -> Bool;
+  fn Less[me: Self](other: U) -> bool;
```
Comment (geoffromer) [inline explorer/data/prelude.carbon]: The changes in this file make sense, but seem unrelated to the topic of this PR. Did they get added by accident?

Comment (pmqtt) [reply explorer/data/prelude.carbon]: Ignore them... This comes from a local merge! Today, I have fixed a bug in the prelude!

File: explorer/interpreter/resolve_names.cpp (hunk 19 above)
Comment (pmqtt) [reply explorer/interpreter/resolve_names.cpp]: Yes, it is a better name!

File: explorer/testdata/for/for-loop-auto.carbon (hunk 20 above)
Comment (pmqtt) [reply explorer/testdata/for/for-loop-auto.carbon]: No! Thanks for this Info!

File: explorer/interpreter/interpreter.cpp (hunk 14 above)
Comment (pmqtt) [reply explorer/interpreter/interpreter.cpp]: Yes, it makes sense!

File: explorer/interpreter/interpreter.cpp (hunk 15 above)
Comment (pmqtt) [reply explorer/interpreter/interpreter.cpp]: Oh you are reight

File: explorer/testdata/for/for-loop-auto.carbon (hunk 21 above)
Comment (pmqtt) [reply explorer/testdata/for/for-loop-auto.carbon]: Yes it is a mistake, should return count

File: explorer/interpreter/interpreter.cpp (hunk 17 above)
Comment (pmqtt) [reply explorer/interpreter/interpreter.cpp]: I can start after this PR! I think this PR is complex enough and we shouldn't take more action. Do you agree?

File: explorer/data/prelude.carbon (hunk 24 above)
Comment (geoffromer) [reply explorer/data/prelude.carbon]: This has now been fixed in trunk, so you can probably merge with that to make these go away.

File: explorer/interpreter/interpreter.cpp [hunk 25]
```diff
@@ -1503,6 +1503,72 @@ auto Interpreter::StepStmt() -> ErrorOr<Success> {
         }
       }
     }
+    case StatementKind::For: {
+      static int CURRENT_INDEX_POS = 1;
+      static int TARGET_VAR_POS = 0;
+      static int END_INDEX_POS = 2;
```
Comment (geoffromer) [inline explorer/interpreter/interpreter.cpp]: The `_pos` suffix on these names could be confusing, because they don't correspond to `act.pos()`. Maybe something like `_RESULT` instead?

Also,
- these should probably be `constexpr`
- we only use all-caps for macros. Constants should [use CamelCase](https://github.com/carbon-language/carbon-lang/blob/trunk/docs/project/cpp_style_guide.md#general-naming-rules).

File: explorer/interpreter/interpreter.cpp [hunk 26]
```diff
@@ -1503,6 +1503,72 @@ auto Interpreter::StepStmt() -> ErrorOr<Success> {
...
+      if (act.pos() == 0) {
+        return todo_.Spawn(
+            std::make_unique<ExpressionAction>(&cast<For>(stmt).loop_target()));
+      }
+      if (act.pos() == 1) {
+        Nonnull<const TupleValue*> source_array =
+            cast<const TupleValue>(act.results().back());
+
+        auto end_index = static_cast<int>(source_array->elements().size());
+        auto start_index = 0;
```
Comment (geoffromer) [inline explorer/interpreter/interpreter.cpp]: I think it would be clearer to delete this variable and use a literal `0` instead.

File: explorer/interpreter/interpreter.cpp [hunk 27]
```diff
@@ -1503,6 +1503,72 @@ auto Interpreter::StepStmt() -> ErrorOr<Success> {
...
+      static int CURRENT_INDEX_POS = 1;
+      static int TARGET_VAR_POS = 0;
+      static int END_INDEX_POS = 2;
+      if (act.pos() == 0) {
+        return todo_.Spawn(
+            std::make_unique<ExpressionAction>(&cast<For>(stmt).loop_target()));
+      }
+      if (act.pos() == 1) {
+        Nonnull<const TupleValue*> source_array =
+            cast<const TupleValue>(act.results().back());
```
Comment (geoffromer) [inline explorer/interpreter/interpreter.cpp]:
```suggestion
            cast<const TupleValue>(act.results()[TARGET_VAR_POS]);
```

File: explorer/interpreter/interpreter.cpp [hunk 28]
```diff
@@ -1503,6 +1503,72 @@ auto Interpreter::StepStmt() -> ErrorOr<Success> {
...
+        auto end_index = static_cast<int>(source_array->elements().size());
+        auto start_index = 0;
+
+        if (start_index < end_index) {
+          act.AddResult(arena_->New<IntValue>(start_index));
+          act.AddResult(arena_->New<IntValue>(end_index));
+          return todo_.Spawn(std::make_unique<PatternAction>(
+              &cast<For>(stmt).variable_declaration()));
+        }
+        return todo_.FinishAction();
```
Comment (geoffromer) [inline explorer/interpreter/interpreter.cpp]: I think it'd be a little clearer to reverse this:

```suggestion
        if (end_index > 0) {
          return todo_.FinishAction();
        }
        act.AddResult(arena_->New<IntValue>(start_index));
        act.AddResult(arena_->New<IntValue>(end_index));
        return todo_.Spawn(std::make_unique<PatternAction>(
            &cast<For>(stmt).variable_declaration()));      
```

That way the general case is handled in "main line" code, and the corner case is handled in a side branch.

File: explorer/interpreter/interpreter.cpp (hunk 17 above)
Comment (geoffromer) [reply explorer/interpreter/interpreter.cpp]: Sure, that works for me, but could you put a TODO comment in the `While` code, to make it clear that the inconsistency is temporary?

File: explorer/interpreter/interpreter.cpp [hunk 29]
```diff
@@ -1503,6 +1503,72 @@ auto Interpreter::StepStmt() -> ErrorOr<Success> {
...
+      }
+      if (act.pos() == 2) {
+        Nonnull<const BindingPlaceholderValue*> loop_var =
+            cast<const BindingPlaceholderValue>(act.results().back());
```
Comment (geoffromer) [inline explorer/interpreter/interpreter.cpp]: I think it'd be clearer to give this result a named index as well.

**Review by geoffromer — APPROVED** [review]
This looks good to me. @zygoloid, do you want to take another look or shall I merge it?

File: explorer/interpreter/resolve_names.cpp [hunk 30]
```diff
@@ -413,6 +413,18 @@ static auto ResolveNames(Statement& statement, StaticScope& enclosing_scope)
...
       break;
     }
+    case StatementKind::For: {
+      StaticScope statement_scope;
+      statement_scope.AddParent(&enclosing_scope);
+      auto& for_stmt = cast<For>(statement);
+      CARBON_RETURN_IF_ERROR(
+          ResolveNames(for_stmt.variable_declaration(), statement_scope));
+      CARBON_RETURN_IF_ERROR(
+          ResolveNames(for_stmt.loop_target(), statement_scope));
```
Comment (zygoloid) [inline explorer/interpreter/resolve_names.cpp]: What should happen if the loop variable is mentioned in the loop target? I think that should be an error. Perhaps we could just reverse these two for now, so the loop variable isn't in scope in the loop target.

```suggestion
      CARBON_RETURN_IF_ERROR(
          ResolveNames(for_stmt.loop_target(), statement_scope));
      CARBON_RETURN_IF_ERROR(
          ResolveNames(for_stmt.variable_declaration(), statement_scope));
```

**Review by zygoloid — APPROVED** [review]
One comment, otherwise looks good.
